runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/FnApiControlClientPoolService.java

```diff
@@ -70,7 +78,11 @@ public static FnApiControlClientPoolService offeringClientsToPool(
   }
 
   @Override
-  public void close() throws Exception {
-    // The clients in the pool are owned by the consumer, which is responsible for closing them
+  public void close() {
+    if (!closed.getAndSet(true)) {
+      for (FnApiControlClient vended : vendedClients) {
```

Comment:
Concurrent race between control() and close()

Thread 1: paused at `FnApiControlClient newClient = FnApiControlClient.forRequestObserver(requestObserver);`
Thread 2: `close()`
Thread 1: Adds new client to `vended`

Also `vended` isn't a concurrent list so the addition caused by Thread 1 during the loop in `close()` could throw a concurrent modification exception.

Reply:
`vended` -> `CopyOnWriteArrayList`, so the last bit is fixed.

I'll update the comment where we add the client, because if the service is being closed we're going to shut down the server, and the client generally should notice that the connection is dead - I could synchronize on the class, but I don't think it's worthwhile, unless you have strong opinions.